taskchampion/src/storage/sqlite.rs

```diff
@@ -1,355 +1,346 @@
+        ];
+        for q in queries {
+            con.execute(q, []).context("Creating table")?;
+        }
+
+        Ok(SqliteStorage { con })
     }
-    fn operations_bucket(&self) -> &'t Bucket<'t, Integer, ValueBuf<Msgpack<Operation>>> {
-        &self.storage.operations_bucket
+}
+
+struct Txn<'t> {
+    txn: Option<rusqlite::Transaction<'t>>,
+}
+
+impl<'t> Txn<'t> {
+    fn get_txn(&self) -> Result<&rusqlite::Transaction<'t>, SqliteError> {
+        self.txn
+            .as_ref()
+            .ok_or(SqliteError::TransactionAlreadyCommitted)
     }
-    fn working_set_bucket(&self) -> &'t Bucket<'t, Integer, ValueBuf<Msgpack<Uuid>>> {
-        &self.storage.working_set_bucket
+
+    fn get_next_working_set_number(&self) -> anyhow::Result<usize> {
+        let t = self.get_txn()?;
+        let result: Option<usize> = t
+            .query_row("SELECT COALESCE(MAX(id), 0) FROM working_set", [], |r| {
+                r.get(0)
+            })
+            .optional()
+            .context("Getting highest working set ID")?;
+
+        Ok(result.unwrap_or(0) + 1)
+    }
+}
+
+impl Storage for SqliteStorage {
+    fn txn<'a>(&'a mut self) -> anyhow::Result<Box<dyn StorageTxn + 'a>> {
+        let txn = self.con.transaction()?;
+        Ok(Box::new(Txn { txn: Some(txn) }))
     }
 }
 
 impl<'t> StorageTxn for Txn<'t> {
     fn get_task(&mut self, uuid: Uuid) -> anyhow::Result<Option<TaskMap>> {
-        let bucket = self.tasks_bucket();
-        let buf = match self.kvtxn().get(bucket, uuid.into()) {
-            Ok(buf) => buf,
-            Err(Error::NotFound) => return Ok(None),
-            Err(e) => return Err(e.into()),
-        };
-        let value = buf.inner()?.to_serde();
-        Ok(Some(value))
+        let t = self.get_txn()?;
+        let result: Option<StoredTaskMap> = t
+            .query_row(
+                "SELECT data FROM tasks WHERE uuid = ? LIMIT 1",
+                [&StoredUuid(uuid)],
+                |r| r.get("data"),
+            )
+            .optional()?;
+
+        // Get task from "stored" wrapper
+        Ok(result.map(|t| t.0))
     }
 
     fn create_task(&mut self, uuid: Uuid) -> anyhow::Result<bool> {
-        let bucket = self.tasks_bucket();
-        let kvtxn = self.kvtxn();
-        match kvtxn.get(bucket, uuid.into()) {
-            Err(Error::NotFound) => {
-                kvtxn.set(bucket, uuid.into(), Msgpack::to_value_buf(TaskMap::new())?)?;
-                Ok(true)
-            }
-            Err(e) => Err(e.into()),
-            Ok(_) => Ok(false),
+        let t = self.get_txn()?;
+        let count: usize = t.query_row(
+            "SELECT count(uuid) FROM tasks WHERE uuid = ?",
+            [&StoredUuid(uuid)],
+            |x| x.get(0),
+        )?;
+        if count > 0 {
+            return Ok(false);
         }
+
+        let data = TaskMap::default();
+        t.execute(
+            "INSERT INTO tasks (uuid, data) VALUES (?, ?)",
```

Comment:
Could maybe just use `INSERT OR IGNORE` here?

Reply:
That would make it hard to have the return-true-if-already-exist return value I think (but I may be misunderstanding?)

Reply:
Ah, that makes sense -- I think postgres has a way to return that information, but figuring that out for sqlite would not be worthwhile.